File: 0113-path-sum-ii/0113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    void helper(List<Integer> sum, List<List<Integer>> ans, TreeNode root, int targetSum)
    {
        if(root == null)
            return;
        sum.add(root.val);
        targetSum -= root.val;
        if(root.left == null && root.right == null && targetSum == 0)
            ans.add(new ArrayList<>(sum));
        helper(sum, ans, root.left, targetSum);
        helper(sum, ans, root.right, targetSum);
        sum.remove(sum.size() - 1);
    }
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>> ans = new ArrayList<>();
        helper(new ArrayList<>(), ans, root, targetSum);
        return ans;
    }
}

File: 0113-path-sum-ii/0113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void helper(vector<int>& sum, vector<vector<int>>& ans, TreeNode* root, int targetSum)
    {
        if(root == NULL)
            return;
        sum.push_back(root->val);
        targetSum -= root->val;
        if(root->left==NULL && root->right == NULL && targetSum == 0)
        {
            ans.push_back(sum);
        }
        helper(sum, ans, root->left, targetSum);
        helper(sum, ans, root->right, targetSum);
        sum.pop_back();
    }
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<vector<int>> ans;
        vector<int> sum;
        helper(sum, ans, root, targetSum);
        return ans;
    }
};